### DISCUSSION TITLE
"Extending" definitions with other properties

### DISCUSSION THREAD

**Discussion by uncenter** [json-schema-org/community#502] [Q&A]
I bet this is a common question but I wasn't able to figure out exactly what I was hoping for. I'm trying to "extend" a defined "type" (something in "$defs") with some other properties:

```json
{
	"$schema": "http://json-schema.org/draft-07/schema#",
	"type": "object",
	"properties": {
		"test": {
			"type": "array",
			"items": {
				"type": "object",
				"allOf": [{ "$ref": "#/$defs/myDefinition" }],
				"properties": {
					"foo": { "type": "string" },
					"bar": { "type": "string" }
				},
				"oneOf": [
					{
						"required": ["foo"]
					},
					{
						"required": ["bar"]
					}
				],
				"additionalProperties": false
			}
		}
	},
	"$defs": {
		"myDefinition": {
			"type": "object",
			"properties": {
				"baz": { "type": "boolean" }
			}
		}
	}
}
```

The idea here is to extend `myDefinition`, which has a single property `baz`, with properties `foo` and `bar` (either `foo` or `bar` is required, and `baz` is optional). In this case, these would be valid.

```json
{
	"test": [{ "foo": "abc" }]
}
```

```json
{
	"test": [{ "foo": "abc", "baz": true }]
}
```

```json
{
	"test": [{ "bar": "abc", "baz": false }]
}
```

The `"allOf": [{ "$ref": "#/$defs/myDefinition" }],`  is from https://stackoverflow.com/a/52579526/17378715, not sure how I should get this to work.

Thanks!

**Comment by gregsdennis** [comment]
There's not a solution for draft 7, but if you can update to draft 2020-12, you'll get the `unevaluatedProperties` keyword to replace `additionalProperties`.

The benefit of `unevaluatedProperties` is that it can "see inside" of subschemas, including through `$ref`s.  `additionalProperties` can only see the immediate subschema, not subschemas that it contains or references.

So `unevaluatedProperties` will pick up `baz` as a defined properties, but `additionalProperties` won't.

You can try this at https://json-everything.net/json-schema.

This passes

```json
{
  "test": [
    {
      "foo": "ad",
      "baz": false
    }
  ]
}
```

This fails

```json
{
  "test": [
    {
      "foo": "ad",
      "bar": "a string",
      "baz": false
    }
  ]
}
```

**Reply by uncenter** [reply]
Thank you so much! I didn't realize I wasn't even on the latest version, that worked amazingly. A quick follow up question though - how can I make this fail with an additional `a` property? I just want the specified properties to be available. 

```json
{
  "test": [
    {
      "foo": "ad",
      "baz": false, 
      "a": true
    }
  ]
}
```

**Reply by plocket** [reply]
To use refs and also prevent additional properties, [this stackoverflow answer](https://stackoverflow.com/a/58798563) worked for me. I had to make sure to use `allOf`.

**Reply by uncenter** [reply]
@plocket what would that look like here?

**Reply by Relequestual** [reply]
Thanks for sharing the stackoverflow answer. That would be right for limited to draft-07, but @gregsdennis provided the ideal solution for 2019-09 and 2020-12. `unevaluatedProperties` was created for this use case.

@uncenter If by "extend" you expect code generation to create classes or types with the knowledge of extensions, you'll likely be out of luck.

Greg also wrote a related article: https://json-schema.org/blog/posts/modelling-inheritance

**Reply by uncenter** [reply]
I don't totally understand what you mean by "code generation to create classes or types with the knowledge of extensions". I'm just hoping to be able to create a reusable definition that I can add additional properties to, while not allowing properties that I haven't actually specified.

Like this in TypeScript:

```ts
type Baz = {
	baz?: string;
};

type Foo = Baz & {
	foo?: string;
	bar?: string;
};

let test: Foo;

// Passing:

test = {
	foo: "foo",
};
test = {
	bar: "bar",
};
test = {
	foo: "foo",
	bar: "bar",
};
test = {
	baz: "baz",
};
test = {
	foo: "foo",
	bar: "bar",
	baz: "string",
};

// Failing:

test = {
	notFooBarBaz: "something",
};
```

**Reply by jdesrosiers** [reply]
> I'm just hoping to be able to create a reusable definition that I can add additional properties to, while not allowing properties that I haven't actually specified.

You can't have one schema that's both extendable and doesn't allow additional properties, but there's a pattern you can use to make it relatively easy to provide both.

`example.schema.json`
```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",

  "type": "object",
  "properties": {
    "baz": { "type": "boolean" }
  },

  "$defs": {
    "closed": {
      "$anchor": "closed",
      "$ref": "#",
      "unevaluatedProperties": false
    }
  }
}
```

With this, the root schema is extendible and the "closed" definition doesn't allow additional properties. If you want the closed version, you can reference `example.schema.json#closed`. If you want to extend it, you can reference `example.schema.json`.

`/extended-example.schema.json`
```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",

  "allOf": [{ "$ref": "example.schema.json" }],
  "properties": {
    "foo": { "type": "string" },
    "bar": { "type": "string" }
  },

  "$defs": {
    "closed": {
      "$anchor": "closed",
      "$ref": "#",
      "unevaluatedProperties": false
    }
  }
}
```

The extended-example can be extended further, but also provides a closed version. Finally, here's an example of using the closed versions of both of those schemas as components to create a closed user-facing schema.

`use-closed-example.schema.json`
```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",

  "type": "object",
  "properties": {
    "example": { "$ref": "example.schema.json#closed" },
    "extended": { "$ref": "extended-example.schema.json#closed" }
  },
  "additionalProperties": false
}
```

**Reply by plocket** [reply]
I know this isn't my issue originally, but I'm having some trouble doing something similar and I'm having trouble parsing this reply. Among other things, I'm not sure how `closed` relates to the rest of it. I'm going to try and make a simpler example of what I think @uncenter is describing combined with the example from @jdesrosiers.

```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "type": "object",
  "properties": {
    "test": {
      "type": "array",
      "items": {
        "type": "object",
        "additionalProperties": false,
        "allOf": [{ "$ref": "#/$defs/myDefinition" }],
        "properties": {
          "foo": { "type": "string" },
          "bar": { "type": "string" }
        },
        "oneOf": [
          { "required": ["foo"] },
          { "required": ["bar"] }
        ]
      }
    }
  },
  "$defs": {
    "myDefinition": {
      "type": "object",
      "unevaluatedProperties": false,
      "properties": {
        "baz": { "type": "boolean" }
      }
    }
  }
}
```

I suspect I have something wrong here, but could we use this as the foundation for the correct way to do it? No external files. Sorry to @uncenter if I've missed the purpose of your code.

**Reply by plocket** [reply]
If you [run this example](https://replit.com/@plocket/pseudo-schema#index.js) in js, I think I (or we) are going for that functionality except using JSONSchema.

**Reply by uncenter** [reply]
@plocket yeah that looks like what I'm going for.

**Reply by plocket** [reply]
I'll go farther and say that [this example](https://replit.com/@plocket/pseudo-schema-2#index.js) is what I'm going for - just having the ref properties acting as if they're just values in the regular `properties` object.

**Reply by uncenter** [reply]
Yep. That's exactly it. I feel like an `extend` keyword could do...

**Reply by gregsdennis** [reply]
@uncenter @plocket please see all of the issues that have been opened in the spec repo about a proposed `$extends` keyword.  Long sorry short: we made `unevaluated*` to solve those problems.

**Reply by uncenter** [reply]
Sure but... right now it really doesn't seem to be solving it!

**Reply by Relequestual** [reply]
In that second example, that's basically the same as what @jdesrosiers showed above.
You can invert it so that the closed version is exposed by default and you have to explicily reference the open version when you need it.

In what way do you feel that doesn't solve the problem?

If it helps, I can modify the single schema you included in [your comment above](https://github.com/orgs/json-schema-org/discussions/502#discussioncomment-7351747) to show how the same principal works. Would that be useful?

**Reply by plocket** [reply]
That would be really useful, @Relequestual, thank you. As minimal a version as is possible, if you can.

**Reply by uncenter** [reply]
Yeah that would be great. Not fully understanding how `jdesrosiers`'s answer helps. Still seems like I can't both extend a schema and prevent additional properties.

**Reply by Relequestual** [reply]
Here it is all in one schema. Generally we advise doing it over multiple files because it makes it simpler and easier to maintain, and you don't have to guess or know about internam names/paths/anchors to get the open or closed version of your thing.

I've included comments in the schema. Please let me know if this doesn't make things clearer.

```JSON
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "type": "object",
  "properties": {
    "test": {
      "type": "array",
      "items": {
        "$comment": "This subschema is now closed, and any property used which isn't defined in applicator, will be invalid/not allowed. You could not use 'additionalProperties' here to do the same thing, because it only considers 'properties' from the same schema object.",
        "type": "object",
        "allOf": [
          { "$ref": "#/$defs/bazExtension" },
          { "$ref": "#/$defs/itemsExtension" }
        ],
        "unevaluatedProperties": false
      }
    }
  },
  "$defs": {
    "bazExtension": {
      "$comment":"This schema is not closed. It can be extended or used as an extension",
      "type": "object",
      "properties": {
        "baz": { "type": "boolean" }
      }
    },
    "bazClosed": {
      "$comment": "This is baz, but closed, so it cannot be extended or used as an extension. (You can do the same for 'itemsExtension' if you wish",
      "$ref": "#/$defs/bazExtension",
      "unevaluatedProperties": true
    },
    "itemsExtension":{
      "$comment":"This schema is not closed. It can be extended or used as an extension",
      "type": "object",
      "properties": {
        "foo": { "type": "string" },
        "bar": { "type": "string" }
      },
      "oneOf": [
        { "required": ["foo"] },
        { "required": ["bar"] }
      ]
    }
  }
}
```

**Reply by plocket** [reply]
This is really clarifying for me, thank you. What is "applicator"? Also, is it also possible to have `properties`, `oneOf`, etc. for the `test.items` object? For example:

```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "type": "object",
  "properties": {
    "test": {
      "type": "array",
      "items": {
        "$comment": "This subschema is now closed, and any property used which isn't defined in applicator, will be invalid/not allowed. You could not use 'additionalProperties' here to do the same thing, because it only considers 'properties' from the same schema object.",
        "type": "object",
        "properties": {
          "foo": { "type": "string" },
          "bar": { "type": "string" }
        },
        "oneOf": [
          { "required": ["foo"] },
          { "required": ["bar"] }
        ],
        "allOf": [
          { "$ref": "#/$defs/bazExtension" }
        ],
        "unevaluatedProperties": false
      }
    }
  },
  "$defs": {
    "bazExtension": {
      "$comment":"This schema is not closed. It can be extended or used as an extension",
      "type": "object",
      "properties": {
        "baz": { "type": "boolean" }
      }
    },
    "bazClosed": {
      "$comment": "This is baz, but closed, so it cannot be extended or used as an extension. (You can do the same for 'itemsExtension' if you wish)",
      "$ref": "#/$defs/bazExtension",
      "unevaluatedProperties": false
    }
  }
}
```

[Fixed value of `/$defs/bazClosed/unevaluatedProperties`, thank you @jdesrosiers]

**Reply by Relequestual** [reply]
Yeah. That's fine.
You're welcome.

Have a read of this article

https://json-schema.org/blog/posts/applicability-json-schema-fundamentals-part-1

**Reply by jdesrosiers** [reply]
There's a typo in the above example. `/$defs/bazClosed/unevaluatedProperties` should be `false`, not `true`.

**Reply by gregsdennis** [reply]
Okay... I got it using `$dynamicRef`.

Base schema:

```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://example.com/baseSchema",
  "type": "object",
  "properties": {
    "test": {
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "baz": { "type": "boolean" }
        }
        "$dynamicRef": "#addons",
        "unevaluatedProperties": false
      }
    }
  },
  "$defs": {
    "defaultAddons": {
      "$dynamicAnchor": "addons",
      "not": true
    }
  }
}
```

Derived schema:

```json

{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://example.com/derived",
  "$defs": {
    "derived": {
      "$dynamicAnchor": "addons",
      "type": "object",
      "properties": {
        "foo": { "type": "string" },
        "bar": { "type": "string" }
      },
      "oneOf": [
        {
          "required": ["foo"]
        },
        {
          "required": ["bar"]
        }
      ]
    }
  },
  "$ref": "baseSchema"
}
```

Here's how it works:

1. For anything validated against just `baseSchema`, the local `$dynamicAnchor` will be resolved, and the `not: true` will ensure that nothing passes.  Everything fails for _only_ this schema.
2. For anything validated against `derived`, _that_ local `$dynamicAnchor` will be resolved, defining new properties.
3. The `unevaluatedProperties` in `baseSchema` can see both the properties defined in `baseSchema` _and_ whatever happens to be defined through the `$dynamicRef`, so it sees
    - `baz` from `baseSchema`
    - `foo` from `derived`
    - `bar` from `derived`

(This may actually be a better approach than what I posted in my inheritance blog post.  Sounds like I need to write another blog post.)

**Reply by plocket** [reply]
Thank you, ~this~ [the conversation before the just previous comment] has clarified a lot for me. Maybe I should make a new discussion for this, but I'm now wondering what `additionalProperties` does that `unevaluatedProperties` can't do. That is, is there a situation in which I should pick the former instead of the latter?

[Later note: I didn't see the post just above this one before writing this. Not a fan of how GitHub hides new posts. I'm a bit confused about what [that previous post] describes, but maybe here isn't the place to go into it.]

**Reply by gregsdennis** [reply]
> I'm now wondering what `additionalProperties` does that `unevaluatedProperties` can't do.

The only difference between the two is what they see:

- `additionalProperties` can see properties defined by sibling keywords.
- `unevaluatedProperties` can see properties defined by sibling keywords _and their subschemas_.

> That is, is there a situation in which I should pick the former instead of the latter?

If you don't want to look at properties defined by subschemas of sibling keywords, then use `additionalProperties`.  I can't think of a scenario when this would be useful off the top of my head, but I'm sure some situation would call for it.

**Reply by gregsdennis** [reply]
As a final addendum, you can change the base addon schema to always pass.  I think this gives a clearer error that it's the `unevaluatedProperties` that's causing the failure rather than the `$dynamic*`.

```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://example.com/baseSchema",
  "type": "object",
  "properties": {
    "test": {
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "baz": { "type": "boolean" }
        }
        "$dynamicRef": "#addons",
        "unevaluatedProperties": false
      }
    }
  },
  "$defs": {
    "defaultAddons": {
      "$dynamicAnchor": "addons"
    }
  }
}
```

Here the `defaultAddons` definition will always pass, which is fine.  Since it doesn't define any new properties, the `unevaluatedProperties` will still fail.

The only thing this _doesn't_ do is support the OO notion of an "is a" relationship.  For example, a schema that represents a `Vehicle` won't validate a `Car`; you'd need to have the `Car` schema.  This is the sticking point for a lot of people, and it's something we still haven't figured out how to model just yet.

**Reply by jdesrosiers** [reply]
There's a typo in the above example. `"$dynamicRef": "addons"` should be `"$dynamicRef": "#addons"`.

**Reply by gregsdennis** [reply]
Ah, yes. I made that correction locally, but it seems it didn't make it over here.

**Reply by Relequestual** [reply]
Also worth noting that "applicators" is [defined in the specification](https://json-schema.org/draft/2020-12/json-schema-core#section-7.5). t sounds like we are getting to the point where reading the specification might be useful, even if all of it doesn't make sense the first read through.
